### PR TITLE
migrate pie chart to echarts

### LINKED ISSUE #27547: Static viz does not respect the pie chart color
**Describe the bug**
We fixed all colors in https://github.com/metabase/metabase/pull/26211, but pie charts

**To Reproduce**
1) Build a question like Orders->Sum of quantity per created_at (month), make a pie chart
2) Add the question to a dashboard, and then do a subscription with it, send a test email.
3) see the pie chart with different colors

**Expected behavior**
Reproduce the same colors in the pie chart

**Screenshots**
image

**Information about your Metabase Installation:**
- Metabase version: 1.45.1

**Severity**
Cosmetic

**Additional context**

How to solve:
Instead of passing `legendColors` to JS donut rendering function provide instance colors (example: waterfall chart). This will make `getColor` prop function return the right instance colors even for EE with reconfigured color settings. Similarly, provide the overridden colors from viz settings. At this point, the static pie chart js code has everything needed to pick the correct segment colors.
After that, implement the pie chart legend in JS instead of the current one in clojure just to make it use the right colors. Use `frontend/src/metabase/static-viz/components/Legend/Legend.tsx` component to do so.

### LINKED ISSUE #33063: Static viz: pie chart omits one category on static viz
### Describe the bug

For this [question](https://stats.metabase.com/question/9684-action-adoption-on-running-instances)

The pie chart omits the "false" number:

On Stats:
image

Static viz:
image

### To Reproduce

1. Go to Stats and open that question
2. See chart
3. Go to our Slack channel with a [subscription](https://metaboat.slack.com/archives/C03H6956HNF/p1690817014232039) 
4. See missing number

### Expected behavior

Should be the same as the frontend chart

### Logs

-

### Information about your Metabase installation

```JSON
Stats
```

### Severity

P2

### Additional context

_No response_

### LINKED ISSUE #33503: Drill through not available on tiny slivers of pie charts
### Describe the bug

When you have a small pie chart wedge, the hover appears, but when you click on the wedge, you do not get the drill in menu when viewing a question.  

https://www.loom.com/share/e8b7eb5305f842bda5652ff815336746?sid=c448bdd1-c75e-4b45-835c-5cecf62c1b1a

### To Reproduce

1. Go the the sample data
2. Join orders to the people table
3. Summarize count by State
4. Filter on the following states: AK, AL, CA, CO, DE, IA, MN, NY, RI, TX, WI
5. Create a pie chart
6. Try to drill in on the DE wedge 

### Expected behavior

If I can click it, I should be able to drill into it

### Logs

_No response_

### Information about your Metabase installation

```JSON
On Master
```

### Severity

It's annoyin

### Additional context

_No response_

### LINKED ISSUE #36096: [Version v47.8] The data display in Pie chart not consistent
### Describe the bug

The data display in Pie chart and and the detail data while we deep dive by move the mouse to the Piechart.

Example: The data of "lack all info" is 94, but the data display here when we move the mouse on is 2282.97
image

### To Reproduce

1. Create data as the capture above
2. Display with pie chart
3. Move the mouse to the pie chart to see the detail data

### Expected behavior

_No response_

### Logs

_No response_

### Information about your Metabase installation

```JSON
- Browser : Chrome
- Version Metabsae : v47.8
```

### Severity

Confuse

### Additional context

_No response_

### LINKED ISSUE #38424: Pie chart % show on chart doesn't show for areas larger than 95%
### Describe the bug

Hello,

For some reason when selecting 'show percentages on the chart' with pie charts areas over 95% in size don't display the percentage

image

vs

image

&

image

I suppose this might be intentional behaviour but at the very least that seems to be undocumented, which caused me many hours of frustration, and ideally would be something we could turn off since having this setting turned on means the percentages will not be in the legend so the information is completely lost otherwise. Also it just makes the chart ugly

I'd consider being able to have them in both the legend and on the chart as a slightly less ideal solution to this too.

### To Reproduce

```
select 'first' as "data", 0 as "amount"

union all

select 'second' as "data", 0 as "amount"

union all

select 'third' as "data", 0 as "amount"

union all

select 'fourth' as "data", 500 as "amount"
```

Turn it into a pie chart visualisation and then tweak the values as you like. This should work regardless of which database you point metabase at.

### Expected behavior

I'd like to be able to configure it to not have a max % cap

### Logs

N/A

### Information about your Metabase installation

```JSON
{
  "browser-info": {
    "language": "en-GB",
    "platform": "Win32",
    "userAgent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/121.0.0.0 Safari/537.36",
    "vendor": "Google Inc."
  },
  "system-info": {
    "file.encoding": "UTF-8",
    "java.runtime.name": "OpenJDK Runtime Environment",
    "java.runtime.version": "11.0.22+7",
    "java.vendor": "Eclipse Adoptium",
    "java.vendor.url": "https://adoptium.net/",
    "java.version": "11.0.22",
    "java.vm.name": "OpenJDK 64-Bit Server VM",
    "java.vm.version": "11.0.22+7",
    "os.name": "Linux",
    "os.version": "5.10.205-195.807.amzn2.x86_64",
    "user.language": "en",
    "user.timezone": "GMT"
  },
  "metabase-info": {
    "databases": [
      "postgres"
    ],
    "hosting-env": "unknown",
    "application-database": "postgres",
    "application-database-details": {
      "database": {
        "name": "PostgreSQL",
        "version": "14.8"
      },
      "jdbc-driver": {
        "name": "PostgreSQL JDBC Driver",
        "version": "42.6.0"
      }
    },
    "run-mode": "prod",
    "version": {
      "date": "?",
      "tag": "vUNKNOWN",
      "hash": "?"
    },
    "settings": {
      "report-timezone": "Europe/London"
    }
  }
}
```

### Severity

Annoying because it's inconsistent and doesn't match other pie charts

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by EmmadUsmani**
Epic https://github.com/metabase/metabase/issues/33281

Closes https://github.com/metabase/metabase/issues/36096
Closes https://github.com/metabase/metabase/issues/33503
Closes https://github.com/metabase/metabase/issues/33063
Closes https://github.com/metabase/metabase/issues/27547
Closes https://github.com/metabase/metabase/issues/38424

### Description

Feature branch for the pie chart migration to echarts.

### How to verify

Can test pie charts locally, or with existing questions from stats on https://metabase-echarts.hosted.staging.metabase.com/

### Demo

https://github.com/user-attachments/assets/2144cb52-0150-4de6-bdf1-104e836e0045

https://github.com/user-attachments/assets/1e998cc9-c517-4055-a50b-83fd89661b25

### Checklist

- [x] Tests have been added/updated to cover changes in this PR

### CODE REVIEW

**Comment by EmmadUsmani** [comment]
* **#43551** <a href="https://app.graphite.dev/github/pr/metabase/metabase/43551?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#43550** <a href="https://app.graphite.dev/github/pr/metabase/metabase/43550?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#43549** <a href="https://app.graphite.dev/github/pr/metabase/metabase/43549?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#43548** <a href="https://app.graphite.dev/github/pr/metabase/metabase/43548?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#43547** <a href="https://app.graphite.dev/github/pr/metabase/metabase/43547?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#43555** <a href="https://app.graphite.dev/github/pr/metabase/metabase/43555?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* `master`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @EmmadUsmani and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: src/metabase/api/pulse.clj
Comment (EmmadUsmani) [inline src/metabase/api/pulse.clj:319]: I'll make sure to do this before merging, but perhaps the reviewer should not approve until I've reverted this change, just in case I forgot

Comment (EmmadUsmani) [reply src/metabase/api/pulse.clj:319]: For context this higher resolution is needed for bug bashing

Comment (adam-james-v) [reply src/metabase/api/pulse.clj:319]: I'm not sure we need to revert that, to be honest. This preview endpoint isn't used for any feature in app I don't think, so it might be ok to just leave the higher resolution.

Comment (EmmadUsmani) [reply src/metabase/api/pulse.clj:319]: I thought it was used for the image previews for slack messages ([example](https://metaboat.slack.com/archives/C03H6956HNF/p1717426968983259)), or does that come from somewhere else?

**Comment by EmmadUsmani** [comment]
Closing this PR in favor of https://github.com/metabase/metabase/pull/43671. It's the same thing but I've renamed the branch from `echarts_pie` to `echarts-pie`

**Comment by EmmadUsmani** [comment]
Reopening as the bug with harbormaster was fixed by https://github.com/metabase/harbormaster/pull/5054, so I'll let the cloud instance keep using the image based on this branch.

File: frontend/src/metabase/visualizations/echarts/pie/model/index.ts
Comment (alxnddr) [inline frontend/src/metabase/visualizations/echarts/pie/model/index.ts:107]: I believe it is still better to keep viz settings related logic in settings definitions. I see there is `"pie._colors"` setting which is computed on fly based on the `settings["pie.colors"]`. Would using it work? If not, we could also define getValue for "pie._colors" and compute colors there so we don't have to do it here

File: frontend/src/metabase/visualizations/echarts/pie/option/index.ts
Comment (alxnddr) [inline frontend/src/metabase/visualizations/echarts/pie/option/index.ts:49]: I'd prefer having `getTotalGraphicOption` to define and return an object instead of cloning `TOTAL_GRAPHIC_OPTION` and then mutating its properties because this approach in my opinion is less readable because now a relatively small object construction is spread in two places and to understand what elements `graphicOption.children[0]` and `graphicOption.children[1]` are I need to go to the definition of `TOTAL_GRAPHIC_OPTION`. Same for `SUNBURST_SERIES_OPTION`

File: frontend/src/metabase/visualizations/echarts/pie/option/index.ts
Comment (alxnddr) [inline frontend/src/metabase/visualizations/echarts/pie/option/index.ts:123]: With the current approach `label` cannot be missing

File: frontend/src/metabase/visualizations/visualizations/PieChart/chart-definition.ts
Comment (alxnddr) [inline frontend/src/metabase/visualizations/visualizations/PieChart/chart-definition.ts:145]: This one seems unused

File: frontend/src/metabase/visualizations/visualizations/PieChart/chart-definition.ts
Comment (alxnddr) [inline frontend/src/metabase/visualizations/visualizations/PieChart/chart-definition.ts:137]: Also currently unused but I wrote another comment about keeping all colors logic here so maybe this will be used

File: src/metabase/api/pulse.clj
Comment (alxnddr) [reply src/metabase/api/pulse.clj:319]: I think it is safe to drop this now since we won't bug bash the chart

**Review by alxnddr — COMMENTED** [review]
Great job, looks way better now 🥇

1) There is a noticeable rendering delay which would be great to eliminate. Probably related to ExplicitSize refresh mode.
https://github.com/user-attachments/assets/2a2a5085-40eb-48c3-8455-5dc3acd47b79

2) It would be great to remove the white border in the dark mode:
Screenshot 2024-07-18 at 9 39 54 PM

File: .loki/reference/chrome_laptop_static_viz_ComboChart_Area_Over_Bar.png
Comment (EmmadUsmani) [reply .loki/reference/chrome_laptop_static_viz_ComboChart_Area_Over_Bar.png:1]: This was an intentional change. In the bug bash we found that the spacing was too small, and I believe we wanted to update this on cartesian charts as well. I'll confirm with Kyle

File: frontend/src/metabase/visualizations/echarts/pie/format.ts
Comment (EmmadUsmani) [reply frontend/src/metabase/visualizations/echarts/pie/format.ts:35]: Maybe not, let me try removing it

File: frontend/src/metabase/visualizations/echarts/pie/option/index.ts
Comment (EmmadUsmani) [reply frontend/src/metabase/visualizations/echarts/pie/option/index.ts:49]: Sure I can change that

File: frontend/src/metabase/visualizations/echarts/pie/model/index.ts
Comment (EmmadUsmani) [reply frontend/src/metabase/visualizations/echarts/pie/model/index.ts:107]: I think having two of the same setting (`pie.colors` and `pie._colors`) is an anti-pattern. It's not clear to developers which one should be used, and bugs will be created by using the wrong setting and not realizing it, or from the two settings diverging in values. We should avoid implementation details that depend on implicit knowledge instead of explicitly clear values in code.

What we should instead do is implement `getValue` for `pie.colors` alone.

File: frontend/src/metabase/visualizations/visualizations/PieChart/chart-definition.ts
Comment (EmmadUsmani) [reply frontend/src/metabase/visualizations/visualizations/PieChart/chart-definition.ts:137]: Will remove

File: frontend/src/metabase/visualizations/visualizations/PieChart/chart-definition.ts
Comment (EmmadUsmani) [reply frontend/src/metabase/visualizations/visualizations/PieChart/chart-definition.ts:145]: Will remove

File: frontend/src/metabase/visualizations/echarts/pie/option/index.ts
Comment (EmmadUsmani) [reply frontend/src/metabase/visualizations/echarts/pie/option/index.ts:123]: That's fine, `label` will always be defined.

**Review by adam-james-v — APPROVED** [review]

File: frontend/src/metabase/visualizations/echarts/pie/format.ts
Comment (EmmadUsmani) [reply frontend/src/metabase/visualizations/echarts/pie/format.ts:35]: Ended up removing it, it had no affect on any formatting

**Comment by EmmadUsmani** [comment]
> Great job, looks way better now 🥇
> 
> 1. There is a noticeable rendering delay which would be great to eliminate. Probably related to ExplicitSize refresh mode.
>    https://github.com/user-attachments/assets/2a2a5085-40eb-48c3-8455-5dc3acd47b79
> 2. It would be great to remove the white border in the dark mode:
> 
> Screenshot 2024-07-18 at 9 39 54 PM

Thanks for the feedback. I've changed the `refreshMode` in `ChartWithLegend` to `debounceLeading`, and confirmed it doesn't adversely effect the region map (only other chart using this component). The delay is shorter although there still is some delay. Given that it's on our product roadmap for v51 to improve legends, we may just want to consolidate legend components at that time, rather than investing more time right now to fix a component that we likely will delete.

File: frontend/src/metabase/visualizations/echarts/pie/model/index.ts
Comment (EmmadUsmani) [reply frontend/src/metabase/visualizations/echarts/pie/model/index.ts]: Ended up moving this logic to `getValue` for `pie.colors`, using the same underlying function both in `chart-definition` and the static viz default computation.

File: frontend/src/metabase/visualizations/echarts/pie/option/index.ts
Comment (EmmadUsmani) [reply frontend/src/metabase/visualizations/echarts/pie/option/index.ts]: Done

File: .loki/reference/chrome_laptop_static_viz_ComboChart_Area_Over_Bar.png
Comment (EmmadUsmani) [reply .loki/reference/chrome_laptop_static_viz_ComboChart_Area_Over_Bar.png:1]: Checked with Maz, there wasn't any feedback against using the same margin so I'll keep it on both charts.

**Review by alxnddr — APPROVED** [review]
🚀

**Pull request merged**
